### PR TITLE
[CHORE] Add stubs and improve comments for pyo3-exposed abstractions, + driveby type/bug fixes.

### PR DESCRIPTION
**Pull request by clarkzinzow**
This PR adds `.pyi` stub definitions for all of our pyo3-exposed classes and functions, adds/improves comments for those abstractions, and makes a bunch of driveby type/bug fixes.

The stub definitions enable stronger mypy type-checking and enable code completion for pyo3 abstractions in IDEs.

### CODE REVIEW

File: daft/daft.pyi [hunk 1]
```diff
@@ -70,4 +73,618 @@ class ImageFormat(Enum):
...
+        """
+        ...
+    def __getitem__(self, idx: int) -> FileInfo: ...
+    def to_table(self) -> PyTable:
+        """
+        Convert to a Daft table with "path", "size", and "num_rows" columns.
+        """
+    def __len__(self) -> int: ...
+
+class S3Config:
```
Comment (samster25) [inline daft/daft.pyi:267]: Should we also fill out the `__init__` for `S3Config`, `AzureConfig`, `GCSConfig`

File: daft/io/_parquet.py [hunk 2]
```diff
@@ -55,6 +56,8 @@ def read_parquet(
 
     file_format_config = FileFormatConfig.from_parquet_config(ParquetSourceConfig())
     if use_native_downloader:
+        if io_config is None:
```
Comment (samster25) [inline daft/io/_parquet.py:59]: Is there an advantage to creating an `io_config` vs leaving it as None?

**Review by samster25 — APPROVED** [review]

File: daft/daft.pyi (hunk 1 above)
Comment (clarkzinzow) [reply daft/daft.pyi:267]: Good catch! I'll add those.

File: daft/io/_parquet.py (hunk 2 above)
Comment (clarkzinzow) [reply daft/io/_parquet.py:59]: Ah, no you're right, we should propagate the `None` in this case.

**Pull request merged**